tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h

```diff
@@ -0,0 +1,365 @@
+  StatusOr<nvinfer1::IConstantLayer*> Constant(const T scalar,
+                                               const int nb_dims) noexcept {
+    TRT_ENSURE(network_);
+    TRT_ENSURE(nb_dims <= nvinfer1::Dims::MAX_DIMS);
+    auto data_type = nvinfer1::DataType::kINT32;
+    if (std::is_floating_point<T>::value) {
+      data_type = nvinfer1::DataType::kFLOAT;
+    }
+    nvinfer1::Dims zero_shape;
+    zero_shape.nbDims = nb_dims;
+    std::fill_n(zero_shape.d, nb_dims, 1);
+    TRT_ShapedWeights const_weights =
+        weight_store_->GetTempWeights(data_type, zero_shape);
+    const_weights.GetPointer<T>()[0] = scalar;
+    nvinfer1::IConstantLayer* const_layer =
+        network_->addConstant(zero_shape, const_weights.GetTrtWeights());
+    TRT_ENSURE(const_layer);
+    return const_layer;
+  };
+
+  // Adds a TensorRT Slice operation to the network.
+  StatusOr<nvinfer1::ISliceLayer*> Slice(
+      nvinfer1::ITensor* input, const nvinfer1::Dims& begin,
+      const nvinfer1::Dims& size, const nvinfer1::Dims& stride) noexcept {
+    TRT_ENSURE(network_);
+    nvinfer1::ISliceLayer* layer =
+        network_->addSlice(*input, begin, size, stride);
+    TRT_ENSURE(layer);
+    return layer;
+  }
+
+  // Adds a TensorRT Shape operation, which determines the runtime shape of the
+  // input tensor, to the network.
+  StatusOr<nvinfer1::IShapeLayer*> Shape(nvinfer1::ITensor* input) {
+    TRT_ENSURE(input);
+    TRT_ENSURE(network_);
+    nvinfer1::IShapeLayer* layer = network_->addShape(*input);
+    TRT_ENSURE(layer);
+    return layer;
+  }
+
+  // Creates a Gather operation on the shape of the input tensor. The output of
+  // the gather operation is a shape tensor of the same size as the input's
+  // shape tensor. The output consists of all zeros except at the specified
+  // indices, where it is equal to the given input tensor's shape in the
+  // corresponding dimension.
+  StatusOr<nvinfer1::IGatherLayer*> GetPartialShapeOf(
+      nvinfer1::ITensor* input, absl::InlinedVector<int64, 4> indices) {
+    TRT_ENSURE(input);
+    TRT_ENSURE(network_);
+    TRT_ENSURE(indices.size() <= nvinfer1::Dims::MAX_DIMS);
+
+    // Get the runtime shape of input;
+    StatusOr<nvinfer1::IShapeLayer*> shape_layer = this->Shape(input);
+    TRT_ENSURE_PTR_OK(shape_layer);
+    nvinfer1::ITensor* runtime_shape = (*shape_layer)->getOutput(0);
+
+    // Create a constant tensor containing the gather indices.
+    // For any dim not in "indices", we mark it size to gather a zero.
+    const int input_nb_dims = input->getDimensions().nbDims;
+    nvinfer1::Dims indices_as_dims;
+    indices_as_dims.nbDims = input_nb_dims;
+    std::fill_n(indices_as_dims.d, input_nb_dims, input_nb_dims);
+    for (auto idx : indices) {
+      TRT_ENSURE(idx < input_nb_dims);
+      indices_as_dims.d[idx] = idx;
+    }
+    StatusOr<nvinfer1::IConstantLayer*> indices_result =
+        this->Constant(indices_as_dims);
```

Comment:
It is understandable that we call "this->Constant" to create a shape tensor for other places, such as the current line 181.
However, it is very confusing that we create a shape tensor here. Can we have a util to just create a 1D INT32 tensor for a list of value and use this and other places where "shape tensor is not relevant"?

Reply:
Fixed. Renamed `Constant(const nvinfer1::Dims&)` to `ConstantShape(const nvinfer1::Dims&)` and added `Constant(const std::vector<int>&)` method